lib/Sema/CSDiagnostics.cpp

```diff
@@ -3775,42 +3756,50 @@ bool PartialApplicationFailure::diagnoseAsError() {
                         ? diag::partial_application_of_function_invalid_swift4
                         : diag::partial_application_of_function_invalid;
 
-  emitDiagnostic(anchor->getNameLoc(), diagnostic, kind);
+  emitDiagnosticAt(anchor->getNameLoc(), diagnostic, kind);
   return true;
 }
 
 bool InvalidDynamicInitOnMetatypeFailure::diagnoseAsError() {
-  auto *anchor = getRawAnchor();
-  emitDiagnostic(anchor->getLoc(), diag::dynamic_construct_class,
+  emitDiagnostic(diag::dynamic_construct_class,
```

Comment:
Should `InvalidDynamicInitOnMetatypeFailure` implement `getLoc` to use the raw anchor's loc? IMO it makes sense to point to the `.init` if it exists.

Reply:
Makes sense, I'll fix both problems by overriding `getAnchor()` on `InvalidInitRefFailure`!